leader-election-impl/src/main/java/com/palantir/leader/proxy/LeadershipStateKeeper.java

```diff
@@ -0,0 +1,160 @@
+ * limitations under the License.
+ */
+
+package com.palantir.leader.proxy;
+
+import com.palantir.common.concurrent.CoalescingSupplier;
+import com.palantir.leader.LeaderElectionService.LeadershipToken;
+import com.palantir.leader.NotCurrentLeaderException;
+import com.palantir.logsafe.SafeArg;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LeadershipStateKeeper<T> {
+    private static final Logger log = LoggerFactory.getLogger(LeadershipStateKeeper.class);
+
+    private final CoalescingSupplier<LeadershipToken> leadershipTokenCoalescingSupplier;
+    private final LeadershipCoordinator leadershipCoordinator;
+    private final AtomicReference<T> delegateRef;
+    private final AtomicReference<LeadershipToken> maybeValidLeadershipTokenRef;
+    private final Supplier<T> delegateSupplier;
+    private volatile boolean isClosed;
+
+    public LeadershipStateKeeper(LeadershipCoordinator leadershipCoordinator, Supplier<T> delegateSupplier) {
+        this.leadershipTokenCoalescingSupplier = new CoalescingSupplier<>(this::getOrUpdateLeadershipToken);
+        this.leadershipCoordinator = leadershipCoordinator;
+        this.delegateSupplier = delegateSupplier;
+        this.delegateRef = new AtomicReference<>();
+        this.maybeValidLeadershipTokenRef = new AtomicReference<>();
+        this.isClosed = false;
+    }
+
+    LeadershipState<T> getLeadershipState() {
+        return ImmutableLeadershipState.<T>builder()
+                .leadershipToken(leadershipTokenCoalescingSupplier.get())
+                .delegate(delegateRef.get())
+                .build();
+    }
+
+    void close() {
+        isClosed = true;
+        clearDelegate();
+    }
+
+    /**
+     * Checks the local ref of leadership token. We need to refresh the delegateRef in case the token has
+     * changed.
+     * This code can be accessed by multiple threads. In order to save all of the requests from locking-unlocking
+     * `this` while trying to update maybeValidLeadershipTokenRef, the invocations are batched using CoalescingSupplier.
+     */
+    @GuardedBy("leadershipTokenCoalescingSupplier")
+    private LeadershipToken getOrUpdateLeadershipToken() {
+        if (!leadershipCoordinator.isStillCurrentToken(maybeValidLeadershipTokenRef.get())) {
+            // we need to clear out existing resources if leadership token has been updated
+            claimResourcesOnLeadershipUpdate();
+            tryToUpdateLeadershipToken();
+        }
+
+        LeadershipToken leadershipToken = maybeValidLeadershipTokenRef.get();
+        if (leadershipToken == null) {
+            // reclaim delegate resources if leadership cannot be gained
+            clearDelegate();
```

Comment:
Forgot to talk about this yesterday, but doing it here can potentially block for a long time. Might be worth it to submit this task to an executor to avoid this

Reply:
This was done, but implemented inefficiently - basically each proxy instance was creating it's own single threaded executor which defeats the purpose of this PR in the first place. 
While correcting that, also spoke with jkong about this, we realized that this could lead to a situation where delegate is refreshed w/o close being called on old delegate instance which is not in line with the past behaviour.
This is why we've decided to undo this change.